metric/instrument/asyncfloat64/asyncfloat64.go

```diff
@@ -0,0 +1,70 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package asyncfloat64 // import "go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
```

Comment:
I think this is not a good package name for the functionality. It does not reflect any "metric" functionality. @rakyll would like your input here, since I know you have good input for these cases.

Reply:
In golang the type name is not very important since majority of the times you don't use that in the assignments or declaration. That is one of the main reason I believe that the split in the package base on the input type does not make sense.

I think something like 'instrument.Float64Counter' and 'instrument.Float64ObserverCounter' is better because of that

Reply:
@bogdandrutu There are 12 instruments in total, (int64, float64) x (6 instruments). This prototype has them in 4 packages, 3 instruments each. How many packages would you propose having?

Reply:
So I created a change that explored this approach.  [Here](https://github.com/MadVikingGod/opentelemetry-go/tree/metrics-api/1-instruments-together).  

There are pros and cons to both approaches.  There is also a takeaway that I will be incorporating into this patch.

The pros for grouping everything into the `instrument` package is that when looking for the kinds of instruments they are all documented together.  This is a nice feature if we can have minimal other things in the package. 
The downside of that approach is that the names become very unwieldy.
What is now a
```
asyncfloat64.UpDownCounter
```
Becomes a 
```
insturment.AsyncFloat64UpDownCounter
```

The other pro is that we could get rid of the namespacing we do, the Meter Interface would then become the 12 Create functions + register.  This was discussed in #2526, and this API in the PR is made this way because of that discussion.  

The one takeaway I have is that the nonrecording implementations can be extracted into a separate package.  That I will incorporate regardless of this discussion.

Reply:
Personally I prefer the one package approach, and that package can either be "metric" where Meter and MeterProvider are defined. 

I also believe that "Sync" can be removed from the names to shorter the common case:

```
+ Meter
  - ObservableInt64Counter(string, ...InstrumentOption) ObservableInt64Counter
  - ObservableInt64UpDownCounter(string, ...InstrumentOption) ObservableInt64UpDownCounter
  - ObservableInt64Gauge(string, ...InstrumentOption) ObservableInt64Gauge

  - ObservableFloat64Counter(string, ...InstrumentOption) ObservableFloat64Counter
  - ObservableFloat64UpDownCounter(string, ...InstrumentOption) ObservableFloat64UpDownCounter
  - ObservableFloat64Gauge(string, ...InstrumentOption) ObservableFloat64Gauge

  - Int64Counter(string, ...InstrumentOption) Int64Counter
  - Int64UpDownCounter(string, ...InstrumentOption) Int64UpDownCounter
  - Int64Histogram(string, ...InstrumentOption) Int64Histogram

  - Float64Counter(string, ...InstrumentOption) Float64Counter
  - Float64UpDownCounter(string, ...InstrumentOption) Float64UpDownCounter
  - Float64Histogram(string, ...InstrumentOption) Float64Histogram
```

Reply:
I slightly prefer the one package approach. End-users can find all `instrument`s in the same place. When the generics feature is ready, we can have intuitive naming for generics functions without creating a new package:

I have no option for Async/Observable naming.

```
+ Meter
  - AsyncCounter(T, ...InstrumentOption) AsyncCounter
  - AsyncUpDownCounter(T, ...InstrumentOption) AsyncUpDownCounter
  - AsyncGauge(T, ...InstrumentOption) AsyncGauge

  - Counter(T, ...InstrumentOption) Counter
  - UpDownCounter(T, ...InstrumentOption) UpDownCounter
  - Histogram(T, ...InstrumentOption) Histogram
```